### ISSUE TITLE
Explain why to version in accept header

### ISSUE THREAD

**Issue by justincy** [interagent/http-api-design#8]
A great example is [Version with Accepts header](https://github.com/interagent/http-api-design#version-with-accepts-header). I wouldn't say it's obvious why putting the version in the accept header is valuable as opposed to versioning in the URL.

**Comment by h2non** [comment]
I had the same question. I'd prefer don't becoming toxic at URL level with the version. Maybe for some particular scenarios could be a reasonable solution, for instance, if you're behind a reverse HTTP proxy or for public APIs (which is particularly more simple to use and differ between versions, as well for end developers).

My vote is to define a new HTTP header, such as `Version` or `X-Version` for that purpose.

**Comment by geemus** [comment]
Yeah, I think the clearer separation provided by having it in headers is nice, but that the format of accept can be a bit cumbersome. Having it more isolated could certainly still keep it in a clear location but avoid some of those pains. If coupled with a query override (kind of like method override) it also provides a path toward specifying in path (via query) where header is inconvenient or not possible.

**Comment by crazytonyi** [comment]
Isn't this what the Pragma header was originally intended for?

The Pragma general-header field is used to include implementation- specific directives that might apply to any recipient along the request/response chain. All pragma directives specify optional behavior from the viewpoint of the protocol; however, some systems MAY require that behavior be consistent with the directives.

**Comment by h2non** [comment]
I don't have a clear about if `Pragma` header should be used for this purpose. According to the [W3C spec](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.32), that header could be filled with  custom values to be used to include implementation specific directives, but historically and empirically no ones places this information in that header, maybe just because it was misunderstood.

I believe that extending HTTP headers with non standards fields is not generally speaking a bad practice, and from the client point of view I think it's simple and explicit defining a custom header for that purpose, and also from my perspective is less error prone.

However, I've to say that multiple alternatives sounds reasonable. I just want to tend to the better one.

**Comment by geemus** [comment]
I suppose Pragma could work, given it's definition, but I've not really seen it used for anything other than disabling caching (and even then, there is a nicer/more specific way to do this in http 1.1). Still, inventing new things has it's own downsides. But I guess to some extent having this as a header all it's own certainly simplifies things vs reusing something else. ie with pragma you might have to stack it on top of some other directives and then you get back to the pain of building/parsing that we currently have with accept. Not that this is, by any means insurmountable (we use accept presently), but that also doesn't mean there couldn't be a clearer, simpler way.

**Comment by mxmlnglt** [comment]
File https://github.com/interagent/http-api-design/blob/master/en/foundations/require-versioning-in-the-accepts-header.md
or https://geemus.gitbooks.io/http-api-design/content/en/foundations/require-versioning-in-the-accepts-header.html
